# Analisi dei dati iniziale e pulizia

In [1]:
import pandas as pd

In [2]:
stock_prices = pd.read_csv('dataset/historical_stock_prices.csv')
stocks = pd.read_csv("dataset/historical_stocks.csv")

## Dataset Historical Stock Prices

In [5]:
stock_prices

ticker   open  close  adj_close    low   high   volume        date
0           AHH  11.50  11.58   8.493155  11.25  11.68  4633900  2013-05-08
1           AHH  11.66  11.55   8.471151  11.50  11.66   275800  2013-05-09
2           AHH  11.55  11.60   8.507822  11.50  11.60   277100  2013-05-10
3           AHH  11.63  11.65   8.544494  11.55  11.65   147400  2013-05-13
4           AHH  11.60  11.53   8.456484  11.50  11.60   184100  2013-05-14
...         ...    ...    ...        ...    ...    ...      ...         ...
20973884    NZF  14.60  14.59  14.590000  14.58  14.62   137500  2018-08-20
20973885    NZF  14.60  14.58  14.580000  14.57  14.61   151200  2018-08-21
20973886    NZF  14.58  14.59  14.590000  14.57  14.63   185400  2018-08-22
20973887    NZF  14.60  14.57  14.570000  14.57  14.64   135600  2018-08-23
20973888    NZF  14.60  14.69  14.690000  14.59  14.69   180900  2018-08-24

[20973889 rows x 8 columns]

In [7]:
stock_prices.isna().sum()

ticker       0
open         0
close        0
adj_close    0
low          0
high         0
volume       0
date         0
dtype: int64

In [8]:
stock_prices.nunique()

ticker          5685
open          807688
close         835181
adj_close    9235753
low           815237
high          821044
volume        385849
date           12274
dtype: int64

In [21]:
stock_prices[stock_prices.duplicated(subset=['ticker','date'])].shape

(0, 8)

Non ci sono righe duplicate per lo stesso (ticker, data)

In [7]:
stock_prices_short = stock_prices.head(500000)

Considero un sottoinsieme del dataset per fare test più rapidi (per verificare la correttezza algoritmica)

In [8]:
stock_prices_short

ticker         open        close   adj_close          low         high  \
0         AHH    11.500000    11.580000    8.493155    11.250000    11.680000   
1         AHH    11.660000    11.550000    8.471151    11.500000    11.660000   
2         AHH    11.550000    11.600000    8.507822    11.500000    11.600000   
3         AHH    11.630000    11.650000    8.544494    11.550000    11.650000   
4         AHH    11.600000    11.530000    8.456484    11.500000    11.600000   
...       ...          ...          ...         ...          ...          ...   
499995    AIG  1332.000000  1203.800049  918.031555  1177.400024  1343.000000   
499996    AIG  1170.000000  1157.000000  882.341309  1090.000000  1197.599976   
499997    AIG  1144.000000  1193.599976  910.252930  1144.000000  1204.000000   
499998    AIG  1160.000000  1154.000000  880.053406  1150.000000  1200.199951   
499999    AIG  1145.199951  1152.000000  878.528259  1124.400024  1157.199951   

         volume        date  
0       4633900  2013-05-08  
1        275800  2013-05-09  
2        277100  2013-05-10  
3        147400  2013-05-13  
4        184100  2013-05-14  
...         ...         ...  
499995  2420000  2004-10-14  
499996  3022800  2004-10-15  
499997  1095900  2004-10-18  
499998   933100  2004-10-19  
499999  1061700  2004-10-20  

[500000 rows x 8 columns]

In [9]:
stock_prices_short.to_csv('dataset/historical_stock_prices_short.csv', index=False)

In [10]:
stock_prices_tiny = stock_prices.head(5)

In [11]:
stock_prices_tiny.to_csv('dataset/historical_stock_prices_tiny.csv', index=False)

## Dataset Historical Stocks

In [6]:
stocks

ticker exchange                                    name  \
0       PIH   NASDAQ  1347 PROPERTY INSURANCE HOLDINGS, INC.   
1     PIHPP   NASDAQ  1347 PROPERTY INSURANCE HOLDINGS, INC.   
2      TURN   NASDAQ                180 DEGREE CAPITAL CORP.   
3      FLWS   NASDAQ                 1-800 FLOWERS.COM, INC.   
4      FCCY   NASDAQ           1ST CONSTITUTION BANCORP (NJ)   
...     ...      ...                                     ...   
6455   ZOES     NYSE                 ZOE&#39;S KITCHEN, INC.   
6456    ZTS     NYSE                             ZOETIS INC.   
6457    ZTO     NYSE               ZTO EXPRESS (CAYMAN) INC.   
6458    ZUO     NYSE                             ZUORA, INC.   
6459   ZYME     NYSE                          ZYMEWORKS INC.   

                 sector                                 industry  
0               FINANCE               PROPERTY-CASUALTY INSURERS  
1               FINANCE               PROPERTY-CASUALTY INSURERS  
2               FINANCE               FINANCE/INVESTORS SERVICES  
3     CONSUMER SERVICES                   OTHER SPECIALTY STORES  
4               FINANCE                     SAVINGS INSTITUTIONS  
...                 ...                                      ...  
6455  CONSUMER SERVICES                              RESTAURANTS  
6456        HEALTH CARE                    MAJOR PHARMACEUTICALS  
6457     TRANSPORTATION        TRUCKING FREIGHT/COURIER SERVICES  
6458         TECHNOLOGY  COMPUTER SOFTWARE: PREPACKAGED SOFTWARE  
6459        HEALTH CARE                    MAJOR PHARMACEUTICALS  

[6460 rows x 5 columns]

In [23]:
stocks[stocks.duplicated(subset=['name'])].shape

(998, 5)

Ci sono 998 valori per la colonna name che sono uguali ad altri nel dataset (duplicati). La colonna name non si può usare come chiave per il job3 (verrà usato il ticker)

In [6]:
stocks.isna().sum()

ticker         0
exchange       0
name           0
sector      1440
industry    1440
dtype: int64

Il dataset che associa i ticker ai settori (stocks) contiene dei settori nulli. Le righe di stock_prices corrispondenti dovranno essere ignorate almeno per il secondo job, che richiede di generare un report per ogni settore.

In [9]:
stocks.nunique()

ticker      6460
exchange       2
name        5462
sector        13
industry     136
dtype: int64

I valori dei ticker sono univoci nel dataset (6460, come lo shape del dataset), quelli dei nomi presentano ripetizioni

In [7]:
stocks['sector'].unique()

array(['FINANCE', 'CONSUMER SERVICES', 'TECHNOLOGY', 'PUBLIC UTILITIES',
       'CAPITAL GOODS', 'BASIC INDUSTRIES', 'HEALTH CARE',
       'CONSUMER DURABLES', nan, 'ENERGY', 'MISCELLANEOUS', 'SECTOR',
       'TRANSPORTATION', 'CONSUMER NON-DURABLES'], dtype=object)

In [16]:
stocks_clean = stocks.loc[(stocks['sector'].notna())]
stocks_clean

ticker exchange                                    name  \
0       PIH   NASDAQ  1347 PROPERTY INSURANCE HOLDINGS, INC.   
1     PIHPP   NASDAQ  1347 PROPERTY INSURANCE HOLDINGS, INC.   
2      TURN   NASDAQ                180 DEGREE CAPITAL CORP.   
3      FLWS   NASDAQ                 1-800 FLOWERS.COM, INC.   
4      FCCY   NASDAQ           1ST CONSTITUTION BANCORP (NJ)   
...     ...      ...                                     ...   
6455   ZOES     NYSE                 ZOE&#39;S KITCHEN, INC.   
6456    ZTS     NYSE                             ZOETIS INC.   
6457    ZTO     NYSE               ZTO EXPRESS (CAYMAN) INC.   
6458    ZUO     NYSE                             ZUORA, INC.   
6459   ZYME     NYSE                          ZYMEWORKS INC.   

                 sector                                 industry  
0               FINANCE               PROPERTY-CASUALTY INSURERS  
1               FINANCE               PROPERTY-CASUALTY INSURERS  
2               FINANCE               FINANCE/INVESTORS SERVICES  
3     CONSUMER SERVICES                   OTHER SPECIALTY STORES  
4               FINANCE                     SAVINGS INSTITUTIONS  
...                 ...                                      ...  
6455  CONSUMER SERVICES                              RESTAURANTS  
6456        HEALTH CARE                    MAJOR PHARMACEUTICALS  
6457     TRANSPORTATION        TRUCKING FREIGHT/COURIER SERVICES  
6458         TECHNOLOGY  COMPUTER SOFTWARE: PREPACKAGED SOFTWARE  
6459        HEALTH CARE                    MAJOR PHARMACEUTICALS  

[5020 rows x 5 columns]

Eliminate le righe con valori di sector nulle dal dataset

In [17]:
stocks_clean.to_csv('dataset/historical_stocks_clean.csv', index=False)